### PR TITLE
KT 14344: Suggest to replace manual range with explicit `indices` call

### PR DESCRIPTION
**Pull request by KilianCallebaut**
[Kotlin Issue](https://youtrack.jetbrains.com/issue/KT-14344)
[Gitlab Issue](https://gitlab.ewi.tudelft.nl/in4315/2018-2019/TI3125TU-swa-20-kotlin/swa-20-kotlin/issues/10)

example:
`0 until collection.size`
and
`0..collection.size-1`

to
`collection.indices`

### CODE REVIEW

File: idea/resources/inspectionDescriptions/ReplacePutWithAssignment.html [hunk 1]
```diff
@@ -1,5 +1,7 @@
 <html>
 <body>
-This inspection reports <b>map.put</b> function calls replaceable with the indexing operator (<b>[]</b>).
+
+This inspection reports <b>until</b> and <b>..</b> operator calls replaceable with the <b>Collection.indices</b>.
```
Comment (devbridie) [inline idea/resources/inspectionDescriptions/ReplacePutWithAssignment.html]: Reset this file, it's not part of the PR

File: idea/resources/inspectionDescriptions/ReplaceManualRangeWithIndicesCallsInspection.html [hunk 2]
```diff
@@ -0,0 +1,5 @@
+<html>
+<body>
+This inspection reports calls to <b>collection.indices</b> instead of <b>RangeTo</b> <b>collection.size-1</b> or <b>Until</b> <b>collection.size</b>
```
Comment (devbridie) [inline idea/resources/inspectionDescriptions/ReplaceManualRangeWithIndicesCallsInspection.html]: it's actually the other way: I like the one you put down here: `This inspection reports <b>until</b> and <b>..</b> operator usages that are replaceable with <b>Collection.indices</b>.`

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        if (expression.getArguments()?.second?.deparenthesize()?.isSizeMinusOne() != true) return
+
+        val constantValue = expression.getArguments()?.first?.constantValueOrNull()
+        val rightValue = (constantValue?.value as? Number)?.toInt() ?: return
+        if (rightValue != 0) return
+
+
+        holder.registerProblem(
+            expression,
+            "'rangeTo or the '..' call should be replaced with .indices call",
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt]: At this point you know it's `rangeTo` so you should remove the `or the '..'` part.

`'rangeTo' usage should be replaced with .indices`

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.calls.callUtil.getType
+import org.jetbrains.kotlin.resolve.constants.ConstantValue
+import org.jetbrains.kotlin.resolve.constants.evaluate.ConstantExpressionEvaluator
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameUnsafe
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode
+
+class ReplaceManualRangeWithIndicesCallsInspection : AbstractKotlinInspection() {
+
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt:29]: Cut down on newlines everywhere

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+    private fun KtExpression.isSizeMinusOne(): Boolean {
+        if (this !is KtBinaryExpression) return false
+        if (operationToken != KtTokens.MINUS) return false
+
+        val leftValue = left?.isSizeOrLength() ?: return false
+        val constantValue = right?.constantValueOrNull()
+        val rightValue = (constantValue?.value as? Number)?.toInt() ?: return false
+        return rightValue == 1 && leftValue
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt:82]: inline `leftValue` here

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
+        val element = descriptor.psiElement as KtExpression
+        val args = element.getArguments() ?: return
+        val second = (args.second as? KtBinaryExpression) ?: return
+        val array = (second.left as? KtDotQualifiedExpression)?.getLeftMostReceiverExpression() ?: return
+
+        element.replace(
+            KtPsiFactory(element).createExpressionByPattern(
+                "$0.indices",
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt]: I'm not sure about this-- we should only replace these calls if the collection receiver actually has `.indices`, and this has never been checked.
An additional check should be made whether or not this size or length call is done on [a type that supports indices](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/indices.html).

File: idea/resources/META-INF/plugin-common.xml [hunk 7]
```diff
@@ -1781,6 +1781,15 @@
                      language="kotlin"
     />
 
+    <localInspection implementationClass="org.jetbrains.kotlin.idea.inspections.ReplaceManualRangeWithIndicesCallsInspection"
+                     displayName="Convert manual ranges to Collection.Indices calls"
```
Comment (devbridie) [inline idea/resources/META-INF/plugin-common.xml]: small `i` here

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+        )
+    }
+
+}
+
+
+private fun KtExpression.deparenthesize() = KtPsiUtil.safeDeparenthesize(this)
+
+private val REGEX_RANGE_TO = """kotlin.(Char|Byte|Short|Int|Long).rangeTo""".toRegex()
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt]: I think maybe we only want to check `Int` here, because `listOf(1,2,3)[4L]` doesn't compile anyways

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+private val REGEX_UNTIL = """kotlin.(Char|Byte|Short|Int|Long).until""".toRegex()
+
+
+fun KtExpression.constantValueOrNull(context: BindingContext? = null): ConstantValue<Any?>? {
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt:126]: Remove this copy and use `AbstractPrimitiveRangeToInspection.constantValueOrNull`

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+        )
+    }
+
+}
+
+
+private fun KtExpression.deparenthesize() = KtPsiUtil.safeDeparenthesize(this)
+
+private val REGEX_RANGE_TO = """kotlin.(Char|Byte|Short|Int|Long).rangeTo""".toRegex()
+private val REGEX_UNTIL = """kotlin.(Char|Byte|Short|Int|Long).until""".toRegex()
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt]: Same as above

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
+
+        return expressionVisitor { expression ->
+            if (expression !is KtBinaryExpression && expression !is KtDotQualifiedExpression) return@expressionVisitor
+
+            val fqName = expression.getCallableDescriptor()?.fqNameUnsafe?.asString() ?: return@expressionVisitor
+
+            if (!fqName.matches(REGEX_RANGE_TO) && !fqName.matches(REGEX_UNTIL)) return@expressionVisitor
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt]: remove this, you're checking it in when

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+            }
+
+        }
+    }
+
+    private fun visitRangeToExpression(expression: KtExpression, holder: ProblemsHolder) {
+        if (expression.getArguments()?.second?.deparenthesize()?.isSizeMinusOne() != true) return
+
+        val constantValue = expression.getArguments()?.first?.constantValueOrNull()
+        val rightValue = (constantValue?.value as? Number)?.toInt() ?: return
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt]: Remove `?: return` here because you're checking != 0 on the line below

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+            ReplaceManualRangeWithIndicesCallQuickFix()
+        )
+
+    }
+
+    private fun visitUntilExpression(expression: KtExpression, holder: ProblemsHolder) {
+        if (expression.getArguments()?.second?.deparenthesize()?.isSizeOrLength() != true) return
+
+        val constantValue = expression.getArguments()?.first?.constantValueOrNull()
+        val rightValue = (constantValue?.value as? Number)?.toInt() ?: return
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt]: Remove ?: return here because you're checking != 0 on the line below

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+class ReplaceManualRangeWithIndicesCallQuickFix : LocalQuickFix {
+    override fun getName() = "Replace with indices"
+
+    override fun getFamilyName() = name
+
+    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
+        val element = descriptor.psiElement as KtExpression
+        val args = element.getArguments() ?: return
+        val second = (args.second as? KtBinaryExpression) ?: return
+        val array = (second.left as? KtDotQualifiedExpression)?.getLeftMostReceiverExpression() ?: return
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt]: Is this correct? It seems to suggest it selects `test` in `test.arr.size`.

Comment (devbridie) [reply idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt]: also the collection is not always an array

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+
+    private fun KtExpression.isSizeMinusOne(): Boolean {
+        if (this !is KtBinaryExpression) return false
+        if (operationToken != KtTokens.MINUS) return false
+
+        val leftValue = left?.isSizeOrLength() ?: return false
+        val constantValue = right?.constantValueOrNull()
+        val rightValue = (constantValue?.value as? Number)?.toInt() ?: return false
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt:81]: Do you need `toInt` here? I think you can use `rightValue == 1`.

**Review by devbridie — CHANGES_REQUESTED** [review]
I think you should also add/change a test so it is demonstrated to work on `listOf` as well.

Otherwise good work! It's a bit harder that I had expected.

File: idea/resources/inspectionDescriptions/ReplacePutWithAssignment.html (hunk 1 above)
Comment (KilianCallebaut) [reply idea/resources/inspectionDescriptions/ReplacePutWithAssignment.html]: Oh snap, seems like I accidentally edited one of my example files

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt (hunk 15 above)
Comment (KilianCallebaut) [reply idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt:81]: Not sure, but took that from another inspection "ReplaceRangetoWithUntilInspection" and it needs to return the same thing

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt (hunk 10 above)
Comment (KilianCallebaut) [reply idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt]: Nah, it came out that wee needed to check for ranges in the .until case. I don't know about what else it can compile to though

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt (hunk 9 above)
Comment (KilianCallebaut) [reply idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt:126]: It doesn't seem like I can reference that

**Review by abansagi — CHANGES_REQUESTED** [review]
The package containing tests is currently named ReplaceManualRangeWithIndicesCallsInspection, however, all the other packages start with a small letter. 😉 

Rest looks good to me!

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt (hunk 5 above)
Comment (KilianCallebaut) [reply idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt:82]: What do you mean by that?

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt (hunk 15 above)
Comment (devbridie) [reply idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt:81]: OK, from ReplaceRangetoWithUntilInspection.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt (hunk 5 above)
Comment (devbridie) [reply idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt:82]: OK, from ReplaceRangetoWithUntilInspection.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt (hunk 9 above)
Comment (devbridie) [reply idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt:126]: Okay, this is complicated because it's defined in a Companion, let's leave it.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt (hunk 10 above)
Comment (devbridie) [reply idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt]: +1

**Review by devbridie — APPROVED** [review]

**Review by abansagi — APPROVED** [review]

**Review by jan-gerling — APPROVED** [review]
Great work Killian!
